### ISSUE TITLE
Clean up query/cluster exceptions/errors

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#4412]
I'm working on resolving merge conflicts between my typed errors branch and `next`.  I've been trying to combine them in a minimally invasive way, which has lead to the following slightly absurd situation:

* The query language throws exceptions of types `exc_t` or `datum_exc_t` which inherit from `base_exc_t` (which inherits from `std::exception`) and have a message and a complicated type attached (LOGIC, RESOURCE, OP_FAILED, etc.).  (`exc_t` also has a backtrace.)
* Most toplevel administration functions called by the ReQL code don't throw an exception; they return a boolean indicating whether they failed and optionally set an `admin_err_t` output variable (this was an `std::string` in `next` before the merge because errors were untyped).  `admin_err_t` has a string and a binary type (FAILED or INDETERMINATE) attached.
  - Usually right after calling these functions, if they failed we produce an `exc_t` or a `datum_exc_t` from the `admin_err_t` right away, so there's a lot of boilerplate.
* Some administrative operations produce an `admin_op_exc_t`, which inherits directly from `std::runtime_error`.  These have a message and a binary type attached.
  - These exceptions are invariably caught and then an `admin_err_t` is assigned to.
* Other administrative operations produce a `no_such_table_exc_t`, an `ambiguous_table_exc_t`, a `failed_table_op_exc_t`, or a `maybe_failed_table_op_exc_t`.  These inherit directly from `std::runtime_error` and come with nothing attached.
  - There are a set of macros in `admin_op_exc.hpp` which automatically catch these excaptions and assign to an `admin_err_t`.
* Some parts of clustering (I think the parts that are involved with a read or a write instead of an administrative operation) produce a `cannot_perform_query_exc_t`, which inherits from `std::exception` and has a message and a binary type attached.
  - This is usually stored directly in a variant or caught and is transformed directly into an `exc_t` or `datum_exc_t`.

Basically, we have 9 custom exception types and a data type which is invariably assigned to by an exception handler and then used to rethrow and exception.  As far as I can tell, administrative operations catch their own exceptions and output errors with an output variable, while all other operations throw exceptions directly.  We spend a truly ridiculous amount of boilerplate code converting between all these different classes (and occasionally discard information; there's at least one place where we catch a `base_exc_t`, throw away the type, output an `admin_err_t` with a fixed type (since `admin_err_t` can't represent the full range of types `base_exc_t` can), and then use the information in the `admin_err_t` to produce an `exc_t`.

There are also some places where we're using exceptions with binary types attached that really seem like they should have complicated types attached (e.g. some of the artificial backends produce what should really be logic errors rather an `OP_FAILED` errors).

---

I think we should instead have two exception types for all these roles: "message + complicated type and "message + complicated type with backtrace".  (Basically `exc_t` and `datum_exc_t` but with `datum_exc_t` broken out so it isn't specific to the `ql` namespace.)  We'd continue to inherit from a common superclass so you can catch both at once if you don't care about the presence or absence of a backtrace.)

I also think we should eliminate the intermediate output variable.  In 95% of cases I've seen while resolving merge conflicts we catch an exception, pack it into the output variable, and then unpack it into another exception that we rethrow on the other side.

**Comment by timmaxw** [comment]
I completely agree about eliminating the intermediate output variable. I think I'd also be fine with unifying `exc_t`, `admin_op_exc_t`, and `cannot_perform_query_exc_t`.

But I think we should keep `no_such_table_exc_t`, `failed_table_op_exc_t`, and `maybe_failed_table_op_exc_t`, for two reasons: First, we often handle them in non-trivial ways rather than letting them bubble up. For example, we sometimes ignore `no_such_table_exc_t`. Second, we don't have the information necessary to produce a useful message at the location where those exceptions are thrown.

In general, I think the new unified `exc_t` should represent a "user-facing exception"; any exception of this type that is thrown should eventually be presented to the user, and we shouldn't do any other processing on it except possibly to add a backtrace or add context to the error message. We shouldn't use it to represent error conditions that we can recover from or take some special action in response to.

**Comment by mlucy** [comment]
@timmaxw -- that sounds reasonable to me.

(My only patch to your description would be to add an exception to "we shouldn't use it to represent error conditions that we can recover from" to allow query-language-level recovery, like e.g. `r.default`.)

**Comment by timmaxw** [comment]
Alright, sounds reasonable.

`ambiguous_table_exc_t` could be handled either way. We could keep it the way it is for consistency with `no_such_table_exc_t`, or we could replace it with `exc_t` because in practice we never actually catch it or process it in any special way.

It might make sense to call the new type `user_exc_t`.

**Comment by mlucy** [comment]
We should do this but it probably won't happen for 2.1.  Bumping to 2.2-polish.
